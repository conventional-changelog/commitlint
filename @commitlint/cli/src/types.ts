export interface CliFlags {
	color: boolean;
	config?: string;
	cwd: string;
	edit?: string | boolean;
	env?: string;
	extends?: (string | number)[];
	help?: boolean;
	'help-url'?: string;
	from?: string;
	'git-log-args'?: string;
	format?: string;
	'parser-preset'?: string;
	quiet: boolean;
	to?: string;
	version?: boolean;
	verbose?: boolean;
	'print-config'?: boolean;
	_: (string | number)[];
	$0: string;
}
