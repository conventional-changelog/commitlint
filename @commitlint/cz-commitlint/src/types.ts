import {RuleConfigCondition, RuleConfigSeverity} from '@commitlint/types';

export type Rule =
	| Readonly<[RuleConfigSeverity.Disabled]>
	| Readonly<[RuleConfigSeverity, RuleConfigCondition]>
	| Readonly<[RuleConfigSeverity, RuleConfigCondition, unknown]>;
