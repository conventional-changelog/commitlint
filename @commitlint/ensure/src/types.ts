export type TargetCaseType =
	| 'camel-case'
	| 'kebab-case'
	| 'snake-case'
	| 'pascal-case'
	| 'start-case'
	| 'upper-case'
	| 'uppercase'
	| 'sentence-case'
	| 'sentencecase'
	| 'lower-case'
	| 'lowercase'
	| 'lowerCase';
