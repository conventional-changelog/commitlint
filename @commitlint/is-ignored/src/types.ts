export type Matcher = (commit: string) => boolean;
