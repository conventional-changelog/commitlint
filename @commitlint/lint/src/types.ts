import {IsIgnoredOptions} from '@commitlint/is-ignored';
import {ParserOptions} from '@commitlint/parse';
import {Rule} from '@commitlint/rules';

export type Linter = (
	commit: string,
	config: LinterRuleConfig,
	options: LinterOptions
) => Promise<LintOutcome>;

export type LinterRuleConfig = {
	[key: string]: any; // todo: use rule configuration from `@commitlint/load`
};

export interface LinterOptions {
	/** If it should ignore the default commit messages (defaults to `true`) */
	defaultIgnores?: IsIgnoredOptions['defaults'];
	/** Additional commits to ignore, defined by ignore matchers  */
	ignores?: IsIgnoredOptions['ignores'];
	/** The parser configuration to use when linting the commit */
	parserOpts?: ParserOptions;

	plugins?: any; // todo: reuse types from `@commitlint/load`
}

export interface LintOutcome {
	/** The linted commit, as string */
	input: string;
	/** If the linted commit is considered valid */
	valid: boolean;
	/** All errors, per rule, for the commit */
	errors: LintRuleOutcome[];
	/** All warnings, per rule, for the commit */
	warnings: LintRuleOutcome[];
}

export interface LintRuleOutcome {
	/** If the commit is considered valid for the rule */
	valid: boolean;
	/** The "severity" of the rule (1 = warning, 2 = error) */
	level: 0 | 1 | 2;
	/** The name of the rule */
	name: string;
	/** The message returned from the rule, if invalid */
	message?: string;
}
