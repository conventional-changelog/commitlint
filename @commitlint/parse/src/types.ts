export interface Commit {
	raw: string;
	header: string;
	type: string | null;
	scope: string | null;
	subject: string | null;
	body: string | null;
	footer: string | null;
	mentions: string[];
	notes: CommitNote[];
	references: CommitReference[];
	revert: any;
	merge: any;
}

export interface CommitNote {
	title: string;
	text: string;
}

export interface CommitReference {
	raw: string;
	prefix: string;
	action: string | null;
	owner: string | null;
	repository: string | null;
	issue: string | null;
}

export type Parser = (
	message: string,
	options: ParserOptions
) => Omit<Commit, 'raw'>;

export interface ParserOptions {
	commentChar?: string;
	headerCorrespondence?: string[];
	headerPattern?: RegExp;
	issuePrefixes?: string[];
	mergeCorrespondence?: string[];
	mergePattern?: RegExp;
	noteKeywords?: string[];
	revertCorrespondence?: string[];
	revertPattern?: RegExp;
}
