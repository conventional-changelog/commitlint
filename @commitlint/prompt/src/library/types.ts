import type {RuleConfigCondition, RuleConfigSeverity} from '@commitlint/types';

export type RuleEntry =
	| [string, Readonly<[RuleConfigSeverity.Disabled]>]
	| [string, Readonly<[RuleConfigSeverity, RuleConfigCondition]>]
	| [string, Readonly<[RuleConfigSeverity, RuleConfigCondition, unknown]>];

export type InputSetting = {
	description?: string;
	enumerables?: Record<
		string,
		{
			description: string;
		}
	>;
	multiline?: boolean;
	header?: {
		length?: number;
	};
};

export type ResultPart = 'type' | 'scope' | 'subject' | 'body' | 'footer';

export type Result = Partial<Record<ResultPart, string | undefined>>;
