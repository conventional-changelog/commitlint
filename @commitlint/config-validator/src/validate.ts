import Ajv from 'ajv';
import {UserConfig} from '@commitlint/types';
import schema from './commitlint.schema.json';
import {formatErrors} from './formatErrors';

const TYPE_OF = [
	'undefined',
	'string',
	'number',
	'object',
	'function',
	'boolean',
	'symbol',
];

export function validateConfig(
	source: string,
	config: unknown
): asserts config is UserConfig {
	const ajv = new Ajv({
		meta: false,
		useDefaults: true,
		validateSchema: false,
		missingRefs: 'ignore',
		verbose: true,
		schemaId: 'auto',
	});

	ajv.addKeyword('typeof', {
		validate: function typeOfFunc(schema: any, data: any) {
			return typeof data === schema;
		},
		metaSchema: {type: 'string', enum: TYPE_OF},
		schema: true,
	});

	const validate = ajv.compile(schema);
	const isValid = validate(config);

	if (!isValid && validate.errors && validate.errors.length) {
		throw new Error(
			`Commitlint configuration in ${source} is invalid:\n${formatErrors(
				validate.errors
			)}`
		);
	}
}
