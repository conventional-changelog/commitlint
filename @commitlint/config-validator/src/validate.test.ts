import {validateConfig} from './validate';
import {UserConfig} from '@commitlint/types';

const validSchemas: Record<string, UserConfig> = {
	empty: {},
	withEmptyExtends: {extends: []},
	withStringExtends: {extends: 'test'},
	withSingleExtends: {extends: ['test']},
	withMultipleExtends: {extends: ['test', 'test2']},
	withFormatter: {formatter: ''},
	withHelpUrl: {helpUrl: ''},
	withRules: {rules: {a: [0], b: [1, 'never'], c: [2, 'never', true]}},
	withParserPresetString: {parserPreset: 'test'},
	withParserPresetObject: {parserPreset: {}},
	withParserPresetObject2: {parserPreset: {name: 'string', path: 'string'}},
	withParserPresetObjectPromise: {
		parserPreset: Promise.resolve({name: 'string'}),
	},
	withParserPresetOpts: {parserPreset: {parserOpts: {test: 1}}},
	withParserPresetOptsPromise: {
		parserPreset: {parserOpts: Promise.resolve({test: 1})},
	},
	withEmptyIgnores: {ignores: []},
	withIgnores: {ignores: [() => true]},
	withDefaultIgnoresTrue: {defaultIgnores: true},
	withDefaultIgnoresFalse: {defaultIgnores: false},
	withEmptyPlugins: {plugins: []},
	withPluginsAsString: {plugins: ['test']},
	withPluginsAsObject: {plugins: [{rules: {}}]},
	shouldSkipAllowAdditionalProperties: {foo: 1},
};

const invalidSchemas: Record<string, any> = {
	whenConfigIsNotObject: [],
	whenConfigIsNotObject2: '',
	extendsAsObject: {extends: {test: 1}},
	extendsWithFunction: {extends: [() => true]},
	formatterAsObject: {formatter: {}},
	helpUrlAsArray: {helpUrl: []},
	rulesAsArray: {rules: ['a']},
	rules1: {rules: {a: [3]}},
	rules2: {rules: {b: [1, 'test', 2, 2]}},
	rules3: {rules: {c: []}},
	rules4: {rules: {d: [[], [], []]}},
	rules5: {rules: {e: {}}},
	parserPreset: {parserPreset: []},
	ignoresFunction: {ignores: () => true},
	ignoresNotFunction: {ignores: [1]},
	defaultIgnoresNotBoolean: {defaultIgnores: 'true'},
	pluginsNotArray: {plugins: 'test'},
	withPluginsAsObject: {plugins: [{}]},
	helpUrlNotString: {helpUrl: {}},
};

describe('validation should pass for', () => {
	test.each(Object.entries(validSchemas))('%s', (file, config) => {
		expect(() => validateConfig(`${file}.js`, config)).not.toThrowError();
	});
});

describe('validation should fail for', () => {
	test.each(Object.entries(invalidSchemas))('%s', (file, config) => {
		expect(() =>
			validateConfig(`${file}.js`, config)
		).toThrowErrorMatchingSnapshot();
	});
});
