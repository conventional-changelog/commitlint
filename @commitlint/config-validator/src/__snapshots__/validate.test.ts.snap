// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`validation should fail for defaultIgnoresNotBoolean 1`] = `
"Commitlint configuration in defaultIgnoresNotBoolean.js is invalid:
	- Property \\"defaultIgnores\\" has the wrong type - should be boolean.
"
`;

exports[`validation should fail for extendsAsObject 1`] = `
"Commitlint configuration in extendsAsObject.js is invalid:
	- Property \\"extends\\" has the wrong type - should be array.
	- Property \\"extends\\" has the wrong type - should be string.
	- \\"extends\\" should match exactly one schema in oneOf. Value: {\\"test\\":1}.
"
`;

exports[`validation should fail for extendsWithFunction 1`] = `
"Commitlint configuration in extendsWithFunction.js is invalid:
	- Property \\"extends[0]\\" has the wrong type - should be string.
	- Property \\"extends\\" has the wrong type - should be string.
	- \\"extends\\" should match exactly one schema in oneOf. Value: [null].
"
`;

exports[`validation should fail for formatterAsObject 1`] = `
"Commitlint configuration in formatterAsObject.js is invalid:
	- Property \\"formatter\\" has the wrong type - should be string.
"
`;

exports[`validation should fail for helpUrlAsArray 1`] = `
"Commitlint configuration in helpUrlAsArray.js is invalid:
	- Property \\"helpUrl\\" has the wrong type - should be string.
"
`;

exports[`validation should fail for helpUrlNotString 1`] = `
"Commitlint configuration in helpUrlNotString.js is invalid:
	- Property \\"helpUrl\\" has the wrong type - should be string.
"
`;

exports[`validation should fail for ignoresFunction 1`] = `
"Commitlint configuration in ignoresFunction.js is invalid:
	- Property \\"ignores\\" has the wrong type - should be array.
"
`;

exports[`validation should fail for ignoresNotFunction 1`] = `
"Commitlint configuration in ignoresNotFunction.js is invalid:
	- \\"ignores[0]\\" should be a function. Value: 1.
"
`;

exports[`validation should fail for parserPreset 1`] = `
"Commitlint configuration in parserPreset.js is invalid:
	- Property \\"parserPreset\\" has the wrong type - should be string.
	- Property \\"parserPreset\\" has the wrong type - should be object.
	- \\"parserPreset\\" should match exactly one schema in oneOf. Value: [].
"
`;

exports[`validation should fail for pluginsNotArray 1`] = `
"Commitlint configuration in pluginsNotArray.js is invalid:
	- Property \\"plugins\\" has the wrong type - should be array.
"
`;

exports[`validation should fail for rules1 1`] = `
"Commitlint configuration in rules1.js is invalid:
	- \\"rules['a'][0]\\" should be equal to one of the allowed values. Value: 3.
	- \\"rules['a']\\" should be a function. Value: [3].
	- \\"rules['a']\\" should match exactly one schema in oneOf. Value: [3].
"
`;

exports[`validation should fail for rules2 1`] = `
"Commitlint configuration in rules2.js is invalid:
	- \\"rules['b']\\" should NOT have more than 3 items. Value: [1,\\"test\\",2,2].
	- \\"rules['b']\\" should be a function. Value: [1,\\"test\\",2,2].
	- \\"rules['b']\\" should match exactly one schema in oneOf. Value: [1,\\"test\\",2,2].
"
`;

exports[`validation should fail for rules3 1`] = `
"Commitlint configuration in rules3.js is invalid:
	- \\"rules['c']\\" should NOT have fewer than 1 items. Value: [].
	- \\"rules['c']\\" should be a function. Value: [].
	- \\"rules['c']\\" should match exactly one schema in oneOf. Value: [].
"
`;

exports[`validation should fail for rules4 1`] = `
"Commitlint configuration in rules4.js is invalid:
	- Property \\"rules['d'][0]\\" has the wrong type - should be number.
	- \\"rules['d'][0]\\" should be equal to one of the allowed values. Value: [].
	- \\"rules['d']\\" should be a function. Value: [[],[],[]].
	- \\"rules['d']\\" should match exactly one schema in oneOf. Value: [[],[],[]].
"
`;

exports[`validation should fail for rules5 1`] = `
"Commitlint configuration in rules5.js is invalid:
	- Property \\"rules['e']\\" has the wrong type - should be array.
	- \\"rules['e']\\" should be a function. Value: {}.
	- \\"rules['e']\\" should match exactly one schema in oneOf. Value: {}.
"
`;

exports[`validation should fail for rulesAsArray 1`] = `
"Commitlint configuration in rulesAsArray.js is invalid:
	- Property \\"rules\\" has the wrong type - should be object.
"
`;

exports[`validation should fail for whenConfigIsNotObject 1`] = `
"Commitlint configuration in whenConfigIsNotObject.js is invalid:
	- Config has the wrong type - should be object.
"
`;

exports[`validation should fail for whenConfigIsNotObject2 1`] = `
"Commitlint configuration in whenConfigIsNotObject2.js is invalid:
	- Config has the wrong type - should be object.
"
`;

exports[`validation should fail for withPluginsAsObject 1`] = `
"Commitlint configuration in withPluginsAsObject.js is invalid:
	- Property \\"plugins[0]\\" has the wrong type - should be string.
	- \\"plugins[0]\\" should have required property '.rules'. Value: {}.
	- \\"plugins[0]\\" should match some schema in anyOf. Value: {}.
"
`;
